### PR TITLE
Remove mixins/ directories from contracts/

### PR DESCRIPTION
**Pull request by jalextowle**
## Description

This change removes the `mixins/` directories from the `contracts/` directory. This change required many parts of the inheritance hierarchy to change. Most of the changes to the inheritance hierarchy occurred in the `contracts/exchange/contracts/src` directory, where `MixinExchangeRichErrors.sol` became a base contract for all of the other `Mixin` files in the directory. 

Another change that is introduced in this pull request is the addition of a `lint:contracts` command in the `package.json` file. This command allows the `contracts/` directory to be linted without linting the rest of the repository. The linting is also done in stages rather than in parallel to make linting easier on machines with weak specs. 

## Testing instructions

1. `yarn build:contracts`
2. `yarn test:contracts`
3. `yarn lint:contracts`

## Types of changes

* New feature (non-breaking change which adds functionality)
* Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [x] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/coordinator/contracts/src/MixinSignatureValidator.sol [hunk 1]
```diff
@@ -19,14 +19,28 @@
...
 
 
 contract MixinSignatureValidator is
-    MSignatureValidator
+    ISignatureValidator
 {
     using LibBytes for bytes;
 
+   // Allowed signature types.
+    enum SignatureType {
```
Comment (hysz) [inline contracts/coordinator/contracts/src/MixinSignatureValidator.sol]: I'd probably also put Enum's in the interface.

File: contracts/exchange/contracts/src/MixinMatchOrders.sol [hunk 2]
```diff
@@ -267,6 +268,48 @@ contract MixinMatchOrders is
         return matchedFillResults;
     }
 
+    function _dispatchTransferFrom(
```
Comment (hysz) [inline contracts/exchange/contracts/src/MixinMatchOrders.sol]: How are these declarations used?

File: contracts/exchange/contracts/src/MixinSignatureValidator.sol [hunk 3]
```diff
@@ -210,6 +231,11 @@ contract MixinSignatureValidator is
         );
     }
 
+    function _getCurrentContextAddress()
```
Comment (hysz) [inline contracts/exchange/contracts/src/MixinSignatureValidator.sol]: Same comment as above - how is this used?

File: contracts/exchange/contracts/src/libs/LibExchangeRichErrorDecoder.sol [hunk 4]
```diff
@@ -18,13 +18,91 @@
...
+import "../MixinSignatureValidator.sol";
+import "../MixinTransactions.sol";
 
 
-contract LibExchangeRichErrorDecoder is
-    MExchangeRichErrorTypes
-{
+contract LibExchangeRichErrorDecoder {
+    // bytes4(keccak256("SignatureError(uint8,bytes32,address,bytes)"))
+    bytes4 internal constant SIGNATURE_ERROR_SELECTOR =
```
Comment (hysz) [inline contracts/exchange/contracts/src/libs/LibExchangeRichErrorDecoder.sol]: See comment above related to these selectors.

File: contracts/exchange/contracts/src/MixinExchangeRichErrors.sol [hunk 5]
```diff
@@ -20,13 +20,113 @@ pragma solidity ^0.5.9;
...
+        INAPPROPRIATE_SIGNATURE_TYPE
+    }
+
+    enum TransactionErrorCodes {
+        NO_REENTRANCY,
+        ALREADY_EXECUTED,
+        EXPIRED
+    }
+
+    // bytes4(keccak256("SignatureError(uint8,bytes32,address,bytes)"))
```
Comment (hysz) [inline contracts/exchange/contracts/src/MixinExchangeRichErrors.sol]: It could be better to put these in a standalone contract. Especially since they're repeated below in `LibExchangeRichErrorDecoder.sol`.

File: contracts/exchange/contracts/src/MixinMatchOrders.sol (hunk 2 above)
Comment (jalextowle) [reply contracts/exchange/contracts/src/MixinMatchOrders.sol]: This (as well as all other abstract functions that I added) are needed to specify the functions in the Mixin. In this case, `_dispatchTransferFrom` and `_getCurrentContextAddress` are used in the private function at the end of the file. 

We could eliminate some of these functions by creating special purpose interfaces for them. I didn't do this initially because it felt like doing this was sort of against the spirit of these changes. This said, I'm happy to try and come up with some interfaces that we could use like that.

File: contracts/exchange/contracts/src/MixinSignatureValidator.sol (hunk 3 above)
Comment (abandeali1) [reply contracts/exchange/contracts/src/MixinSignatureValidator.sol]: I think it probably makes sense to just inherit `MixinTransactions` whenever they are used.

File: contracts/exchange/contracts/src/MixinWrapperFunctions.sol [hunk 6]
```diff
@@ -451,4 +452,17 @@ contract MixinWrapperFunctions is
         }
         return fillResults;
     }
+
+    function _fillOrder(
```
Comment (abandeali1) [inline contracts/exchange/contracts/src/MixinWrapperFunctions.sol]: This contract should probably just inherit `MixinExchangeCore` as well.

File: contracts/exchange/contracts/src/interfaces/IDispatcher.sol [hunk 7]
```diff
@@ -0,0 +1,13 @@
+pragma solidity ^0.5.9;
+
+
+contract IDispatcher {
```
Comment (abandeali1) [inline contracts/exchange/contracts/src/interfaces/IDispatcher.sol]: Why do we need this? Interfaces should only contain external/public functions.

File: contracts/exchange/contracts/src/interfaces/IExchangeCore.sol [hunk 8]
```diff
@@ -21,10 +21,45 @@ pragma experimental ABIEncoderV2;
 
 import "@0x/contracts-exchange-libs/contracts/src/LibOrder.sol";
 import "@0x/contracts-exchange-libs/contracts/src/LibFillResults.sol";
+import "@0x/contracts-exchange-libs/contracts/src/LibOrder.sol";
```
Comment (abandeali1) [inline contracts/exchange/contracts/src/interfaces/IExchangeCore.sol]: Why is this duplicated?

File: contracts/exchange/contracts/src/interfaces/IExchangeCore.sol [hunk 9]
```diff
@@ -57,4 +92,14 @@ contract IExchangeCore {
         public
         view
         returns (LibOrder.OrderInfo memory orderInfo);
+
+    function _isValidOrderWithHashSignature(
```
Comment (abandeali1) [inline contracts/exchange/contracts/src/interfaces/IExchangeCore.sol]: Why does this internal function need to be in the interface?

File: contracts/exchange/contracts/src/interfaces/IExchangeRichErrors.sol [hunk 10]
```diff
@@ -1,30 +1,12 @@
...
-
-*/
-
 pragma solidity ^0.5.9;
 
-import "@0x/contracts-utils/contracts/src/mixins/MRichErrorTypes.sol";
-import "@0x/contracts-exchange-libs/contracts/src/LibOrder.sol";
 
+contract IExchangeRichErrors {
+    enum AssetProxyDispatchErrorCodes {
```
Comment (abandeali1) [inline contracts/exchange/contracts/src/interfaces/IExchangeRichErrors.sol]: Styling nit: Add a newline after `contract IExchangeRichErrors {` (I've seen this in a few other places as well).

File: contracts/exchange/contracts/src/libs/LibExchangeRichErrorDecoder.sol [hunk 11]
```diff
@@ -34,14 +35,14 @@ contract LibExchangeRichErrorDecoder is
         public
         pure
         returns (
-            SignatureErrorCodes errorCode,
+            IExchangeRichErrors.SignatureErrorCodes errorCode,
```
Comment (abandeali1) [inline contracts/exchange/contracts/src/libs/LibExchangeRichErrorDecoder.sol]: Why this change?

File: contracts/exchange/contracts/src/interfaces/IFillAssertions.sol [hunk 12]
```diff
@@ -0,0 +1,35 @@
+pragma solidity ^0.5.9;
+
+import "@0x/contracts-exchange-libs/contracts/src/LibOrder.sol";
+import "@0x/contracts-exchange-libs/contracts/src/LibFillResults.sol";
+
+
+contract IFillAssertions {
```
Comment (abandeali1) [inline contracts/exchange/contracts/src/interfaces/IFillAssertions.sol]: I think I'd prefer removing this and `IGetContext` in favor of just inheriting `MixinExchangeCore`. Down the line the actual implementations for these functions can potentially live in a separate mixin. I'd leave it as is for now, since I think we'll be modifying the structure of these contracts a good amount soon.

File: contracts/exchange/contracts/src/MixinMatchOrders.sol (hunk 2 above)
Comment (hysz) [reply contracts/exchange/contracts/src/MixinMatchOrders.sol]: Ohh, I see what's going on. So declarations in `MFoobar` would need to be replaced by their respective implementation, defined in `MixinFoobar`. In this case, `MixinMatchOrders` would now inherit from `MixinAssetProxyDispatcher` rather than `MAssetProxyDispatcher`.

File: contracts/exchange/contracts/src/interfaces/IExchangeRichErrors.sol (hunk 10 above)
Comment (jalextowle) [reply contracts/exchange/contracts/src/interfaces/IExchangeRichErrors.sol]: Do we need to put a new line after every contract declaration, or only when the declaration fits on a singe line?

Ex. Would the following need a new line after `{`?

```
contract A is 
    B,
    C
{
    enum D { }
}

File: contracts/utils/contracts/src/RichErrors.sol [hunk 13]
```diff
@@ -18,12 +18,12 @@
 
 pragma solidity ^0.5.9;
 
-import "./mixins/MRichErrors.sol";
 
+contract RichErrors {
+    // bytes4(keccak256("Error(string)"))
```
Comment (abandeali1) [inline contracts/utils/contracts/src/RichErrors.sol]: Nit: add newline before this comment

File: contracts/exchange/contracts/src/MixinTransactions.sol [hunk 14]
```diff
@@ -54,20 +53,20 @@ contract MixinTransactions is
     }
 
     /// @dev Executes a batch of Exchange method calls in the context of signer(s).
-    /// @param transactions Array of 0x transactions containing salt, signerAddress, and data.
+    /// @param zeroxTransactions Array of 0x transactions containing salt, signerAddress, and data.
```
[the remark this answers is not in the document]
Comment (jalextowle) [reply contracts/exchange/contracts/src/MixinTransactions.sol]: It was shadowing the mapping, so I changed it to avoid the compiler warning. I can switch it back if you'd rather.

File: contracts/exchange/contracts/src/interfaces/IExchangeRichErrors.sol (hunk 10 above)
Comment (abandeali1) [reply contracts/exchange/contracts/src/interfaces/IExchangeRichErrors.sol]: Only after the declaration fits on a single line (so you wouldn't need one in your example).

File: contracts/exchange/contracts/src/interfaces/IExchangeCore.sol (hunk 8 above)
Comment (abandeali1) [inline contracts/exchange/contracts/src/interfaces/IExchangeCore.sol]: Hmm, looks like the extra import is still here.
